Planning time limit

skrzypo987 opened the pull request.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:
```diff
@@ -61,6 +61,7 @@
     private String queryExecutionPolicy = "all-at-once";
     private Duration queryMaxRunTime = new Duration(100, TimeUnit.DAYS);
     private Duration queryMaxExecutionTime = new Duration(100, TimeUnit.DAYS);
+    private Duration queryMaxPlanningTime = new Duration(10, TimeUnit.MINUTES);
```
Is this limit reasonable?

hashhar replied:
How does this differ/relate to the optimizer timeout config? Does `queryMaxPlanningTime` include time from admission until split creation?

10 mins seems like a very generous planning timeout. FWIW, you will hit optimizer timeout after ~3min anyway (default).

findepi replied:
Is this limit safe to introduce? i.e. can it break some legitimate workloads that work correctly today?

cc @martint @electrum @phd3

sopel39 replied:
Maybe we should make it infinite by default?

martint replied:
10 minutes is plenty of time. It's extremely unlikely for any well-behaving system to be taking that long to plan.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/execution/DataDefinitionExecution.java:
```diff
@@ -219,6 +220,12 @@ public void fail(Throwable cause)
+    @Override
+    public boolean isPlanningFinished()
+    {
+        return getState().ordinal() > PLANNING.ordinal();
```
There are 4 methods that look exactly like that. One way of fixing that is introducing `getState` method to `TrackedQuery` interface.

sopel39 replied:
Add method to `QueryState` (`isPlanningFinished`). I would also use explicit boolean in `QueryState` (`planningFinished`) or explicit case `io.trino.execution.StageState#canScheduleMoreTasks`.

findepi commented on testing/trino-tests/src/test/java/io/trino/execution/resourcegroups/db/TestQueuesDb.java:
```diff
@@ -306,6 +330,26 @@ public void testQueryExecutionTimeLimit()
+                queryRunner,
+                testSessionBuilder()
+                        .setCatalog("tpch")
+                        .setSchema("sf100000")
```
Does it matter?

skrzypo987 replied:
no

findepi commented on testing/trino-tests/src/test/java/io/trino/execution/resourcegroups/db/TestQueuesDb.java:
```diff
@@ -306,6 +330,26 @@ public void testQueryExecutionTimeLimit()
+                        .setSource("dashboard")
+                        .setSystemProperty(QUERY_MAX_PLANNING_TIME, "1ms")
+                        .build(),
+                LONG_PLANNING_QUERY);
```
Use Mock connector with some operation invoked during planning doing a sleep.

Also, planning time limit doesnt belong to TestQueues, does it?

skrzypo987 commented:
Added a test that freezes during planning. To my amazement it got interrupted properly. Changed docs so that setting this limit will not guarantee instant termination of the query
Also changed the default to 100 days. Let's keep it backward-compatible and move the default value discussion to subsequent PR.

findepi commented:
> move the default value discussion to subsequent PR.


do you plan to make it?

skrzypo987 commented:
> do you plan to make it?

Sure, it is easy. Just need to open a PR and propose an absurd default value. Then a lot of of Trino seniors will come arguing about the actual value and settle with something.
(Disclaimer: That was a joke)

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:
```diff
@@ -284,6 +285,19 @@ public QueryManagerConfig setQueryMaxExecutionTime(Duration queryMaxExecutionTim
         return this;
     }
 
+    @NotNull
```
specify min duration `@MinDuration`

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryState.java:
```diff
@@ -73,4 +73,22 @@ public boolean isDone()
+            case FAILED:
+                return true;
+        }
+        return true;
```
this should throw an exception

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/DataDefinitionExecution.java:
```diff
@@ -219,12 +219,6 @@ public void fail(Throwable cause)
         stateMachine.transitionToFailed(cause);
```
change commit text,
```
Expose query state in TrackedQuery.
```
`isDone` method is no longer exposed

sopel39 commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,92 @@
+import static java.lang.Integer.MAX_VALUE;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+// Tests relies on timeouts so additional query runners may disturb
```
what did you mean here?

skrzypo987 replied:
The test needs to finish between 5s and 10s time window. Any task running in the background may increase chance of flakiness.

sopel39 replied:
I think the comment is not complete

sopel39 commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,92 @@
+            Thread.sleep(MAX_VALUE);
+        }
+        catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
set a future on interrupted exception and validate that it was set in `testInterruptFetchingStatistics` after planning interrupted

sopel39 replied:
What about future there?

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:
```diff
@@ -801,6 +804,12 @@ public boolean transitionToDispatching()
+        this.planningThread = planningThread;
+        return transitionToPlanning();
+    }
+
```
remote the other method?

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:
```diff
@@ -906,6 +915,10 @@ public boolean transitionToFailed(Throwable throwable)
             }
         }
 
+        if (oldState == PLANNING && planningThread != null) {
```
what about `transitionToCanceled`

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:
```diff
@@ -801,6 +804,12 @@ public boolean transitionToDispatching()
         return queryState.setIf(DISPATCHING, currentState -> currentState.ordinal() < DISPATCHING.ordinal());
     }
 
+    public boolean transitionToPlanning(Thread planningThread)
```
actually, could we handle interruption in `SqlQueryExecution` (via state change listener)?

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -378,6 +382,12 @@ public void start()
                 }
+                Thread planningThread = currentThread();
+                stateMachine.getStateChange(PLANNING).addListener(() -> {
+                    if (stateMachine.getQueryState() == FAILED) {
```
Does cancellation always transition to failed state?

skrzypo987 replied:
Both `transitionToCancelled` and `transitionToFailed` set state to `FAILED`

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -84,14 +84,18 @@
 import static com.google.common.base.Preconditions.checkArgument;
```
this commit should go before `Add query planning time limit`

sopel39 commented on docs/src/main/sphinx/admin/properties-query-management.rst:
```diff
@@ -13,6 +13,18 @@ The maximum allowed time for a query to be actively executing on the
+* **Session property:** ``query_max_planning_time``
+
+The maximum allowed time for a query to be actively planning the execution.
+After this period the coordinator will make its best effort to stop the
```
maybe just `to stop the query`. It's kind of odd that coordinator is `actively planning the execution`  and then we `stop the query execution`

sopel39 commented on docs/src/main/sphinx/admin/properties-query-management.rst:
```diff
@@ -13,6 +13,18 @@ The maximum allowed time for a query to be actively executing on the
 cluster, before it is terminated. Compared to the run time below, execution
 time does not include analysis, query planning or wait times in a queue.
 
+``query.max-planning-time``
```
please review @mosabua

m57lyra replied:
LGTM.

sopel39 commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,92 @@
+            {
+                return ImmutableList.of(MockConnectorFactory.builder()
+                        .withGetColumns(ignored -> ImmutableList.of(new ColumnMetadata("name", VARCHAR)))
+                        // Apply Join happens inside optimizer so this should model most blocking tasks in planning phase
```
nit: use something simpler maybe, like predicate  or projection pushdown?

sopel39 commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -36,6 +37,8 @@
 {
     private DistributedQueryRunner queryRunner;
 
+    private volatile boolean interrupted;
```
swap commits order and make this change part of commit that adds test

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -378,6 +382,12 @@ public void start()
+                Thread planningThread = currentThread();
+                stateMachine.getStateChange(PLANNING).addListener(() -> {
+                    if (stateMachine.getQueryState() == FAILED) {
+                        planningThread.interrupt();
```
This might not be a planning thread anymore (when planning is finished) when query goes to `FAILED` state. In fact, this executor thread might be running planning for some other query. I think interruption needs to run in critical section:
1. planning thread is setting `planningThread=null` when planning is finished
2. other thread is calling `interrupt`

skrzypo987 replied:
The listener will fire only when the state changes from `PLANNING` to `FAILED`. Once planning is complete the state changes immediately to `STARTING` and the thread will not get interrupted.

skrzypo987 replied:
Ok, you are right, the thread might be reused. I will sync it.

skrzypo987 replied:
Fixed. PTAL

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:
```diff
@@ -284,6 +285,20 @@ public QueryManagerConfig setQueryMaxExecutionTime(Duration queryMaxExecutionTim
     }
 
+    @NotNull
+    @MinDuration("1ns")
```
nit: why not 0? Duration itself can be `0`.

skrzypo987 replied:
Removed the constraint to be consistent with `max-execution-time`

sopel39 commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -261,6 +262,11 @@ public SystemSessionProperties(
+                durationProperty(
+                        QUERY_MAX_PLANNING_TIME,
+                        "Maximum planning time of a query",
+                        queryManagerConfig.getQueryMaxPlanningTime(),
```
please add validation here too (`>=`).

skrzypo987 replied:
`Duration` already validates the negative numbers

sopel39 commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,98 @@
+    {
+        assertThatThrownBy(() -> queryRunner.execute("SELECT * FROM t1 WHERE col = 'abc'"))
+                .hasMessageContaining("Query exceeded the maximum planning time limit of 5.00s");
+        assertThat(interrupted).isTrue();
```
This is racy, planning thread might have hiccup and set `interrupted = true` later than `queryRunner.execute` returns.

Use https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html instead.

sopel39 reviewed (approved):
lgtm % comments

sopel39 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -379,12 +387,27 @@ public void start()
                 planDistribution(plan);
 
+                synchronized (this) {
+                    planningThread = null;
```
you need to unset interrupted flag here too. Fortunately `Threadh.interrupt` and `Thread.interrupted` are in happens-before relationship

findepi commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -379,12 +387,27 @@ public void start()
                 registerDynamicFilteringQuery(plan);
                 planDistribution(plan);
 
+                synchronized (this) {
```
add `@GuardedBy` annotation to the field

findepi commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -379,12 +387,27 @@ public void start()
+                    if (stateMachine.getQueryState() == FAILED) {
+                        synchronized (this) {
+                            if (planningThread != null) {
+                                planningThread.interrupt();
```
How does this relate to the interrupt delivered here
https://github.com/trinodb/trino/blob/f13283bcc5a7507fa5396cdec9afbda93d642ee6/core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java#L102 
?

skrzypo987 replied:
It is a different thread. `queryExecutionFuture.cancel(true)` interrupts thread started in `LocalDispatchQueryFactory::createDispatchQuery`.
The actual planning thread is started from that thread in `LocalDispatchQuery::startExecution`

skrzypo987 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -129,6 +134,9 @@
     private final DynamicFilterService dynamicFilterService;
 
+    @Nullable
+    private Thread planningThread;
```
It is used in async lambdas and cannot be a variable. So I could either decorate it in some `AtomicReference` type of object of make it a field.

skrzypo987 replied:
It is being set to `null` later on

martint replied:
It's not clear to me why we need that. The only reason would be to prevent the wrong thread from being interrupted, but that can't really happen. If the query transitions to FAILED during planning, it *must* still be executing the `start()` method, since all the planning methods are synchronous. It is only then that the thread is interrupted, which means it will be the right thread.

I can't think of a combination of actions that would result in `start()` completing and the query not already be in a state other than PLANNING, which means the interruption wouldn't happen beyond that point.

findepi replied:
if we remove `planningThread`, we also remove synchronization around setting/unsetting it. Wouldn't that introduce race conditions? Would we be able to still reason that `start` didnt compete yet?

martint replied:
The synchronization doesn't prevent any race conditions -- it's only for JMM semantics regarding updating/reading that field from multiple threads.

In any case, I think it's worth adding a comment along the lines of what I described above. Reasoning about concurrency is hard, and we don't have the right primitives here to make it easier.

skrzypo987 replied:
This was implemented to prevent the interruption in an unlikely case of the thread being already reused for another query. This can be simulated by placing long sleep just before the `interrupt()` method. That way we can interrupt the next query executed in this thread.

sopel39 replied:
State change listener can execute any time in the future (even when `planningThread` started serving another query)

martint replied:
Fair enough. The use of `directExecutor()` in the addListener call threw me off.

However, if can happen *only* if the planning fails from the call to `fail()` in the catch at the bottom of this method.

Still, I would write this code as follows to keep the state and logic self-contained and not have to reason about a class-level field, it's lifetime beyond this method (or other methods, etc). Ideally, we should be able to remove listeners, but there's no facility to do so:

```java
final AtomicReference<Thread> planningThread = new AtomicReference<>(currentThread());
stateMachine.getStateChange(PLANNING).addListener(() -> {
    if (stateMachine.getQueryState() == FAILED) {
        Thread thread = planningThread.get();
        if (thread != null) {
            thread.interrupt();
        }
    }
}, directExecutor());

try {
    PlanRoot plan = planQuery();
    // DynamicFilterService needs plan for query to be registered.
    // Query should be registered before dynamic filter suppliers are requested in distribution planning.
    registerDynamicFilteringQuery(plan);
    planDistribution(plan);
}
finally {
    planningThread.set(null);

    // clear the interrupted flag in case there was a race condition where
    // the planning thread was interrupted right after planning completes above
    Thread.interrupted();
}
```

sopel39 replied:
> However, if can happen only if the planning fails from the call to fail() in the catch at the bottom of this method.

Why only then? Query might be interrupted via timeout and `thread.interrupt();` might be delivered in arbitrary time after that.

Other than that keeping logic self-contained within method makes sense

martint replied:
> Why only then? Query might be interrupted via timeout and thread.interrupt(); might be delivered in arbitrary time after that.

Ah yes, scratch that. I was thinking of the case where the interrupt causes those methods to throw.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -379,12 +387,27 @@ public void start()
                 }
 
+                planningThread = currentThread();
+                stateMachine.getStateChange(PLANNING).addListener(() -> {
```
This is for the case when the query was failed explicitly by invoking `fail` method. So the planning continues and we want to make the best effort to stop it.

martint replied:
Can you show me an example of where that happens? If the query is marked as failed during planning, planning should abort on its own.

skrzypo987 replied:
The last commit `QueryTracker::enforceTimeLimits`. Just failing the query will not stop the planning thread, especially the hanging IO

sopel39 reviewed (commented):
Did you push you changes?

findepi commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -561,7 +561,6 @@ public void fail(Throwable cause)
     }
 
-    @Override
     public boolean isDone()
```
no longer public?

findepi commented on core/trino-main/src/main/java/io/trino/execution/QueryState.java:
```diff
@@ -73,4 +73,23 @@ public boolean isDone()
+            case FINISHED:
+            case FAILED:
+                return true;
+            default:
```
https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md#avoid-default-clause-in-exhaustive-enum-based-switch-statements

findepi commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -135,6 +136,7 @@
     private final DynamicFilterService dynamicFilterService;
 
     @Nullable
+    @GuardedBy("this")
```
belongs to prev cmt

findepi commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -405,6 +407,11 @@ public void start()
+                    if (Thread.interrupted()) {
+                        // Clears interrupted flag in case this thread is reused.
+                        // The query will fail to continue because of the 'failed' state.
+                        log.debug("Planning thread has been interrupted");
```
Belongs to prev cmt.

findepi commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,101 @@
+import static io.trino.SystemSessionProperties.QUERY_MAX_PLANNING_TIME;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.Integer.MAX_VALUE;
```
Don't static import since there are many constants with this name in various classes.

findepi commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,101 @@
+        queryRunner.createCatalog("mock", "mock");
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown()
+    {
+        queryRunner.close();
+        queryRunner = null;
+    }
```
You can instead extend from `AbstractTestQueryFramework` which will handle cleanup for you.

findepi reviewed (commented):
(just skimming)

skrzypo987 commented:
@martint PTAL if the current state is ok for you

skrzypo987 commented:
@martint I used the `planningTime` from `QueryStateTimer` to check the timeout. One commit less and not using `executionStartTime`. PTAL

martint commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,94 @@
+        Session defaultSession = testSessionBuilder()
+                .setCatalog("mock")
+                .setSchema("default")
+                .setSystemProperty(QUERY_MAX_PLANNING_TIME, "5s")
```
We need to make this much tighter for this test or come up with a better way to test this.

skrzypo987 replied:
What exactly do you mean?
The feature is about making best effort to stop the planning, so ensuring it works in some cases should be enough IMO

martint replied:
I'm saying that 5s is to long to sleep in a test.

martint commented on testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java:
```diff
@@ -0,0 +1,94 @@
+    private <T> T freeze()
+    {
+        try {
+            Thread.sleep(Integer.MAX_VALUE);
```
Pick a more reasonable value. A broken implementation will, effectively, make this thread hang forever in the test suite and hold on to resources beyond the execution of this test. Ideally, when the test completes, this should be "released" regardless of success or failure, so it might be better to simulate this freeze with a CountdownLatch that gets released when the test is torn down.

skrzypo987 replied:
used CoundDownLatch

martint commented on core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:
```diff
@@ -379,12 +395,32 @@ public void start()
                 planDistribution(plan);
 
+                synchronized (this) {
+                    if (Thread.interrupted()) {
```
This shouldn't be necessary. It should be taken care of by the thread pool.

sopel39 replied:
This thread does not go into pool immediately. This might not be a problem, but it seems cleaner to clear interrupted flag since we interrupt planning specifically. wdyt?

martint replied:
If planning is interrupted before this line, the above calls (planQuery, registerDynamicFilteringQuery, planDistribution) will throw and this will not be reached. 

The only case where it can happen is if there is a race condition where the planning completes just about the time the timeout kicks in and interrupts the thread right after `planDistribution` returns or while it's in some part that doesn't get affected by the interrupt so it doesn't throw. It definitely deserves a comment, and the debug statement is not very useful. Also see my comment above on how I'd restructure this section.

sopel39 replied:
> If planning is interrupted before this line, the above calls (planQuery, registerDynamicFilteringQuery, planDistribution) will throw and this will not be reached.

That will only happen if these methods will throw interrupted exception on IO which might not be the case.

> Also see my comment above on how I'd restructure this section.

+1

martint replied:
> That will only happen if these methods will throw interrupted exception on IO which might not be the case.

That should not happen unless the code can do something reasonable with a lack of result from that IO. If we're silently ignoring those interruptions and making results up, that's a bigger problem.

martint reviewed (commented):
Looks good, but let's change the default max planning time to something more reasonable (e.g., 10 minutes). 100 days is meaningless -- we might as well make it optional and default to no limit specified.

sopel39 commented:
mind automation

skrzypo987 commented:
There are some problems with GHA at the moment, I will restart it later
